### PR TITLE
Ramil Gauss

### PR DESCRIPTION
**Pull request by RamilGauss**
Issue: https://github.com/boostorg/dll/issues/32

"DeImangle does not match when smart_library calls import_Imangled with const T* or std ::string #32"

### CODE REVIEW

File: include/boost/dll/detail/demangling/itanium.hpp [hunk 1]
```diff
@@ -184,7 +184,7 @@ std::string mangled_storage_impl::get_mem_fn(const std::string &name) const
              '(' + parser::arg_list(*this, func_type()) + ')'
              + const_rule<Class>() + volatile_rule<Class>();
 
-    auto found = std::find_if(storage_.begin(), storage_.end(), [&](const entry& e) {return e.demangled == matcher;});
+    auto found = std::find_if(storage_.begin(), storage_.end(), [&](const entry& e) {return e.demangled.find( matcher ) != std::string::npos;});
```
Comment (apolukhin) [inline include/boost/dll/detail/demangling/itanium.hpp]: Please add a comment `// Linux export table contains int MyClass::Func<float>(), but expected in import_mangled MyClass::Func<float>() without returned type.`

The fix could give incorrect results for Other::MyClass::Func<float>() and MyClass::Func<float>(). Looks like the check should be close to the following `e.demangled == matcher || e.demangled.find(" " + matcher) != std::string::npos;`

File: test/template_method_linux_test.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// For only Linux testing
```
Comment (apolukhin) [inline test/template_method_linux_test.cpp]: Add a copyright note with tour name and a Boost license. Here's an example:

https://github.com/boostorg/dll/blob/9049914417137bee972a52e8d09dfa4e2d0252f1/include/boost/dll/import.hpp#L1-L7

File: test/template_method_linux_test.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <boost/dll/runtime_symbol_info.hpp>
+#include <boost/dll/smart_library.hpp>
+#include <boost/dll/import_mangled.hpp>
+#include <boost/dll/import_class.hpp>
+#include <iostream>
+#include <memory>
+
```
Comment (apolukhin) [inline test/template_method_linux_test.cpp]: I'll register the test in build files. Jamfiles are hard and creepy, you do not have to deal with them

**Comment by apolukhin** [comment]
Many thanks for the fix and test!

**Comment by XiaLiChao82** [comment]
I added a code in the itanium.h file that under the GCC 7.3.0 compiler can resolve the type name of the pointer and the type name of the template as generated by the typeid() operator, but I don't have that many compiler versions, how do I need to run the test case?




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Antony Polukhin"<notifications@github.com&gt;; 
发送时间: 2019年11月28日(星期四) 凌晨3:52
收件人: "boostorg/dll"<dll@noreply.github.com&gt;; 
抄送: "Subscribed"<subscribed@noreply.github.com&gt;; 
主题: Re: [boostorg/dll] Ramil Gauss (#35)




Many thanks for the fix and test!

**Comment by XiaLiChao82** [comment]
I added a code in the itanium.h file that under the GCC 7.3.0 compiler can parse the type name of the pointer and the type name of the template to match the type name generated by the typeid() operator, but I don't have that many compiler versions, how do I need to run the test case?




------------------ 原始邮件 ------------------
发件人: "Antony Polukhin"<notifications@github.com&gt;; 
发送时间: 2019年11月28日(星期四) 凌晨3:52
收件人: "boostorg/dll"<dll@noreply.github.com&gt;; 
抄送: "Subscribed"<subscribed@noreply.github.com&gt;; 
主题: Re: [boostorg/dll] Ramil Gauss (#35)




Many thanks for the fix and test!

**Comment by apolukhin** [comment]
Just make a Pull Request with some tests for one platform. I'll try to take care of the rest.

**Comment by XiaLiChao82** [comment]
OK



---Original---

**Comment by XiaLiChao82** [comment]
Hi,I have submitted the pull request.



---Original---

**Pull request merged**